Full project with tests

h1067675 opened the pull request.

h1067675 committed:
Basic project code without tests

h1067675 committed:
Проект с тестами

h1067675 committed:
Change go version

h1067675 committed:
Изменил версию go

h1067675 committed:
Вернул версию 1.23

h1067675 committed:
Check routing

h1067675 committed:
Checked

h1067675 committed:
checked context

h1067675 committed:
Added logs and automatic authorization after registration

h1067675 committed:
Added more logs

h1067675 committed:
Added logs for loader

h1067675 committed:
Checking http client

h1067675 committed:
issue with the configurator has been resolved and some logs have been added

h1067675 committed:
fixed json parsing and order status update

h1067675 committed:
fixed json struct

h1067675 committed:
added some more logs

h1067675 committed:
added some more logs

h1067675 committed:
fixed db transaction

h1067675 committed:
minor edits

h1067675 committed:
minor changes in routing

h1067675 committed:
Final edits?

h1067675 committed:
And now?

h1067675 committed:
Разбираюсь почему ругается на версию го

h1067675 committed:
downgraded the go version

h1067675 committed:
everything is checked

rktkov commented on internal/configurer/configurer.go:
```diff
@@ -0,0 +1,163 @@
+	"github.com/h1067675/gophermart/internal/logger"
+)
+
+type Configurer interface {
```
Объявление интерфейса лучше делать в месте использования, а не реализации

h1067675 replied:
Исправлено.
В целом так и делал, но почему-то посчитал, что так будет уместнее.

h1067675 committed:
Полностью изменен процесс обработки заказа:
1. Заказ получен от пользователя и записан в базу данных.
2. Заказ помещается в буферизованный канал с новыми заказами.
3. Обработчик новых заказов проверяет свободна ли очередь выполнения запросов к системе расчета бонусов и если есть свободные места в канале то отправляет в работу, если нет, то отправляет в кучу с максимальным приоритетом для дальнейшей обработки из кучи.

При запуске сервера запускается loader который проверяет заказы с неконечными статусами в базе данных и если такие есть отправляет их в кучу для дальнейшей обработки, парралельно запускается процесс который отправляет заказы из кучи в работу.

Всего работает 10 воркеров которые парралельно делают запросы к серверу расчета бонусов и при получании неокончательного статуса отправляются в кучу для дальнейшей обработки постепенно понижая приоритет. При получении статуса 429 пакет клиета делает останавливает запросы к данному серверу и возвращает их в работу только после истечения таймера.

h1067675 committed:
Так же появилась мысль перевести очередь в Kafka, но пока не получилось разобраться с запуском и настройкой, поэтому пока очередь реализована при помощи кучи.

+ Мелкие правки для прохождения vettest

h1067675 committed:
+ исправил объявление интерфейса

h1067675 committed:
Добавлены схемы

h1067675 committed:
Схемы добавлены в readme

Perederey commented on cmd/loader/loader.go:
```diff
@@ -0,0 +1,341 @@
+	}
+}
+
+func (l *Loader) NStartLoader() {
```
переименуй NStartLoader -> StartLoaderWorkers

h1067675 replied:
Исправлено

Perederey commented on internal/compress/compress.go:
```diff
@@ -0,0 +1,33 @@
+	return w.Writer.Write(b)
+}
+
+func CompressHandler(next http.Handler) http.Handler {
```
Хорошая реализация middleware

h1067675 replied:
Все строго по урокам

Perederey commented on internal/authorization/authorization.go:
```diff
@@ -0,0 +1,70 @@
+	UserAuthorization(login string, password string) (userID int, err error)
+}
+
+const secretKey = "mysecretkey"
```
Не храни секретный ключ в коде, используй переменные окружения

h1067675 replied:
Добавил переменную окружения SECRET_KEY

Perederey commented on cmd/accrual/main.go:
```diff
@@ -0,0 +1,173 @@
+	return c.Router
+}
+
+func (c *Connect) StartServer() error {
```
Вот тут используй контекст с таймаутом вместо пустого контекста в методе StartServer

h1067675 replied:
Удалил вообще, так как он ошибочно остался, даже не помню для чего его хотел использовать

Perederey commented on cmd/client/client.go:
```diff
@@ -0,0 +1,125 @@
+	"time"
+)
+
+type Client struct {
```
Создай интерфейс HTTPClient для мокирования в тестах

Perederey commented on cmd/accrual/main.go:
```diff
@@ -0,0 +1,173 @@
+)
+
+// General structure
+type Connect struct {
```
Переименуй Connect в Server

h1067675 replied:
Готово

Perederey commented on cmd/accrual/main.go:
```diff
@@ -0,0 +1,173 @@
+	return limiter
+}
+
+func (c *Connect) limit(next http.Handler) http.Handler {
```
Функцию limit можно упростить, вынеся логику в отдельный метод

h1067675 replied:
Вынес в отдельную функцию логику запроса лимита подключений для пользователя, однако код функции не сильно упростился, либо нужно убирать все в отдельную функцию, но какой в этом смыл?

Perederey commented on cmd/gophermart/handlers.go:
```diff
@@ -0,0 +1,329 @@
+}
+
+// POST /api/user/balance/withdraw
+func (c *Connect) UserGetBalanceWithdrawHandler(response http.ResponseWriter, request *http.Request) {
```
 можно вынести валидацию в отдельную функцию, типа такого 

```go
func validateWithdrawalRequest(body []byte) (*withdrawal, error) {
    var w withdrawal
    if err := json.Unmarshal(body, &w); err != nil {
        return nil, err
    }
    order, err := strconv.Atoi(w.Order)
    if err != nil {
        return nil, fmt.Errorf("invalid order number")
    }
    if !luhn.Valid(order) {
        return nil, fmt.Errorf("invalid luhn number")
    }
    return &w, nil
}```

h1067675 replied:
Сделано

Perederey commented on cmd/depository/accrual.go:
```diff
@@ -0,0 +1,46 @@
+	"github.com/h1067675/gophermart/internal/logger"
+)
+
+type accrualAnswer struct {
```
переименуй структуру в AccrualResponse - так будет понятнее и соответствует правилам Go

h1067675 replied:
готово

Perederey commented on cmd/loader/loader.go:
```diff
@@ -0,0 +1,341 @@
+	q.QueryTicker--
+}
+
+func (l *Loader) NgetOrderStatusFromServerAPI(order Order) (result Order, err error) {
```
Метод слишком большой, лучше разбить на подфункции

h1067675 replied:
готово

Perederey reviewed (approved):
Привет! Код в целом хороший, но есть несколько моментов, которые стоит улучшить:
- Нужно упростить сложные функции, разбив их на более мелкие
- нейминг


Продолжай в том же духе!

h1067675 committed:
Исправлен замечания
